File: coin-change/coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        
        
        dp = [float('inf')] * (amount+1)
        dp[0] = 0
        
        for denom in coins:
            
            for i in range(denom, amount+1):
                
                remain = i - denom
                dp[i] = min (dp[remain] + 1, dp[i])
        
        return dp[amount] if dp[amount] != float('inf') else -1
        

File: coin-change/coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        
        
        int[] dp = new int[amount+1];
        Arrays.fill(dp, Integer.MAX_VALUE - 1);
        dp[0] = 0;
        
        for(int denom : coins){
            for(int i = denom; i < amount+1; i++){
                int remain = i - denom;
                dp[i] = Math.min(dp[remain] + 1, dp[i]);
            }
        }
        
        
        
        
        return dp[amount] != Integer.MAX_VALUE - 1 ? dp[amount] : -1;
        
        
    }
}